src/api/java/com/minecolonies/api/entity/combat/combat/ThreatTable.java

```diff
@@ -0,0 +1,257 @@
+package com.minecolonies.api.entity.combat.combat;
+
+import net.minecraft.entity.LivingEntity;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.minecolonies.api.util.constant.Constants.TICKS_SECOND;
+
+/**
+ * Threat table class, basically a list of entities with an associated threat value
+ */
+public class ThreatTable
+{
+    /**
+     * Melee range
+     */
+    private static final int MELEE_RANGE = 8;
+
+    /**
+     * Max tracking time
+     */
+    private static final int MAX_TRACKING_TICKS = TICKS_SECOND * 120;
+
+    /**
+     * List holding the threat entries
+     */
+    private List<ThreatTableEntry> threatList = new ArrayList<>();
+
+    /**
+     * Currently targeted entity
+     */
+    private int currentTargetIndex = 0;
+
+    /**
+     * The owner entity of this table
+     */
+    private final LivingEntity owner;
+
+    public ThreatTable(final LivingEntity owner)
+    {
+        this.owner = owner;
+    }
+
+    /**
+     * Adds X threat to the given entity
+     *
+     * @param attacker entity to add the value to
+     * @param threat   threat value to add
+     */
+    public void addThreat(final LivingEntity attacker, final int threat)
+    {
+        ThreatTableEntry threatTableEntry = null;
+        int index = threatList.size();
+
+        for (int i = 0; i < index; i++)
+        {
+            final ThreatTableEntry entry = threatList.get(i);
+            if (entry.getEntity() == attacker)
+            {
+                threatTableEntry = entry;
+                index = i;
+                break;
+            }
+        }
+
+        if (threatTableEntry == null)
+        {
+            threatTableEntry = new ThreatTableEntry(attacker);
+            threatList.add(threatTableEntry);
+            threatTableEntry.addThreat(Math.max(0, 10 - (owner.blockPosition().distManhattan(attacker.blockPosition()) / 4)));
+        }
+
+        threatTableEntry.addThreat(threat);
+        adaptTableToThreat(index);
+    }
+
+    /**
+     * Get the threat value for the given entity
+     *
+     * @param attacker entity
+     * @return threat value
+     */
+    public int getThreatFor(final LivingEntity attacker)
+    {
+        ThreatTableEntry threatTableEntry = null;
+        int index = threatList.size();
+
+        for (int i = 0; i < index; i++)
+        {
+            final ThreatTableEntry entry = threatList.get(i);
+            if (entry.getEntity() == attacker)
+            {
+                return entry.getThreat();
+            }
+        }
+
+        return 0;
+    }
+
+    /**
+     * Resorts the given index after its threat value changed
+     *
+     * @param index
+     */
+    private void adaptTableToThreat(final int index)
```

Comment:
Couldn't we use an "OrderedList" that adapts itself after add?

Reply:
ye didnt find any nicer alternative with stepwise sorting